src/index/addrindex.cpp

```diff
@@ -0,0 +1,263 @@
+    unsigned int seed;
+
+    std::unique_ptr<CDBIterator> iter(NewIterator());
+    std::pair<char, uint8_t> key;
+
+    // If key is in the index already, read it and return.
+    iter->Seek(seed_key);
+    if (iter->Valid() && iter->GetKey(key) && key == seed_key && iter->GetValue(seed)) {
+        return seed;
+    }
+
+    // Generate a random key and write it to the index.
+    seed = GetRandInt(std::numeric_limits<int>::max());
+    Write(seed_key, seed);
+    return seed;
+}
+
+AddrIndex::~AddrIndex() {}
+
+unsigned int AddrIndex::GetAddrId(const CScript& script) {
+    std::vector<unsigned char> script_data;
+    for (auto it = script.begin(); it != script.end(); ++it) {
+        script_data.push_back(*it);
+    }
+    return MurmurHash3(m_hash_seed, script_data);
+}
+
+bool AddrIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CBlockUndo block_undo;
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    std::vector<std::pair<DBKey, DBValue>> entries;
+
+    const bool not_genesis_block = (pindex->nHeight > 0);
+    if (not_genesis_block && !UndoReadFromDisk(block_undo, pindex)) {
+      return false;
+    }
+
+    for (size_t i = 0; i < block.vtx.size(); ++i) {
+        const CTransaction& tx = *(block.vtx[i]);
+        const uint256 tx_hash = tx.GetHash();
+        for (size_t j = 0; j < tx.vout.size(); ++j) {
+            CScript script_pub_key = tx.vout[j].scriptPubKey;
+            DBKey key(DBKeyType::CREATED, GetAddrId(script_pub_key), COutPoint(tx_hash, j));
+            entries.emplace_back(key, std::make_pair(pos, script_pub_key));
+        }
+
+        // Skip coinbase inputs.
+        if (not_genesis_block && i > 0) {
+            const CTxUndo& tx_undo = block_undo.vtxundo[i-1];
+            for (size_t k = 0; k < tx.vin.size(); ++k) {
+                CScript spent_outputs_scriptpubkey = tx_undo.vprevout[k].out.scriptPubKey;
+                DBKey key(DBKeyType::SPENT, GetAddrId(spent_outputs_scriptpubkey), tx.vin[k].prevout);
+                entries.emplace_back(key, std::make_pair(pos, spent_outputs_scriptpubkey));
+            }
+        }
+        pos.nTxOffset += ::GetSerializeSize(tx, CLIENT_VERSION);
+    }
+
+    return m_db->WriteToIndex(entries);
+}
+
+bool AddrIndex::DB::WriteToIndex(const std::vector<std::pair<DBKey, DBValue>> &entries)
+{
+    CDBBatch batch(*this);
+    for (const auto& entry : entries) {
+        batch.Write(entry.first, entry.second);
+    }
+    return WriteBatch(batch);
+}
+
+// FindTxsByScript fills the spends_result vector with outpoints corresponding
+// to the output spent with the given script, and the transaction it was spent
+// in. creations_result is filled with outpoints for outputs created with this
+// script as their script pubkey, and the transactions they were created in.
+bool AddrIndex::FindTxsByScript(const CScript& script,
```

Comment:
Shouldn't this method take the count as a parameter and return early when enough entries are found?